lib/polyfill/pip.js

```diff
@@ -0,0 +1,245 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+goog.provide('shaka.polyfill.PiP');
+
+goog.require('shaka.log');
+goog.require('shaka.polyfill.register');
+
+
+/**
+ * @namespace shaka.polyfill.PiP
+ *
+ * @summary A polyfill to provide PiP support in Safari.
+ */
+
+
+/**
+ * Install the polyfill if needed.
+ *
+ * @this {HTMLMediaElement}
+ */
+shaka.polyfill.PiP.install = function() {
+  if (!window.HTMLVideoElement) {
+    // Avoid errors on very old browsers.
+    return;
+  }
+
+  let proto = HTMLVideoElement.prototype;
+  if (proto.requestPictureInPicture &&
+    document.exitPictureInPicture) {
+    // No polyfill needed.
+    return;
+  }
+
+  if (proto.webkitSupportsPresentationMode &&
+    typeof proto.webkitSetPresentationMode === 'function') {
+    shaka.log.debug('PiP.install');
+
+    /**
+     * polyfill document.pictureInPictureElement
+     */
+    Object.defineProperty(document, 'pictureInPictureElement', {
+      get() {
+        if (!document.polyfillPictureInPictureElement) {
+          const videoElementList = document.querySelectorAll('video');
```

Comment:
Another place where we might want to generalize to cover audio.

Also, why do we search for the elements instead of always using the cached value?

Do we need to handle PiP being triggered outside of this polyfill?  Is it possible that an element is already in PiP mode when the polyfill is installed?  If these answers are probably "no", then we could drastically simplify this.